# ◆Backtest
###### Create Date：2020/02/14　Author：M.Hasegawa
### ────────────────────────────────────────────────────────────────

#### 【table of contents】
- 0.[**Import module**](#Import_module)
- 1.[**Import data**](#Import_data)
- 2.[**Strategy**](#Strategy)
- 3.[**Backtest**](#Backtest)
- 4.[**Optimisation**](#Optimisation)
- 5.[**Output**](#Output)
- 6.[**Appendix.VBA**](#Vba)

#### ※reference
- data : [OANDA](http://developer.oanda.com/rest-live/rates/)
- tool：[Backtesting.py](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting)
- sample code1 : https://kernc.github.io/backtesting.py/doc/examples/Multiple%20Time%20Frames.html
- sample code2 : https://saidataisei.hatenablog.com/entry/2019/10/13/003622
- sample code3 : http://mmorley.hatenablog.com/entry/fx_backtesting01


## 0. Import module<a id='Import_module'></a>
- pip install backtesting
- pip install oandapyV20
- pip install git+https://github.com/oanda/oandapy.git
- pip install TA_Lib-0.4.17-cp37-cp37m-win_amd64.whl

###### TA-LIB
- ref: https://qiita.com/ConnieWild/items/cb50f36425a683c914d2
- ref: https://www.lfd.uci.edu/~gohlke/pythonlibs/#ta-lib

In [87]:
%matplotlib inline
import oandapy
import pytz
import configparser
import pandas as pd
import numpy  as np
import talib as ta
from datetime         import datetime
from oandapyV20       import API
from backtesting      import Backtest, Strategy
from backtesting.lib  import crossover
from backtesting.test import SMA, GOOG

config = configparser.ConfigParser()
config.read('./Settings/config_OANDA.txt')
account_id = config['oanda']['account_id']
api_key = config['oanda']['api_key']

import warnings
warnings.filterwarnings("ignore")

# ============================================================================
# Conv Japan Time
# ============================================================================
def iso_to_jp(iso):
    date = None
    try:
        date = datetime.strptime(iso, '%Y-%m-%dT%H:%M:%S.%fZ')
        date = pytz.utc.localize(date).astimezone(pytz.timezone("Asia/Tokyo"))
    except ValueError:
        try:
            date = datetime.strptime(iso, '%Y-%m-%dT%H:%M:%S.%f%z')
            date = date.astimezone(pytz.timezone("Asia/Tokyo"))
        except ValueError:
            pass
    return date

# ============================================================================
# Conv Format
# ============================================================================
def date_to_str(date):
    if date is None:
        return ''
    return date.strftime('%Y/%m/%d %H:%M:%S')

# ============================================================================
# SMA
# ============================================================================
def SMA(array, n):
    return pd.Series(array).rolling(n).mean()

# ============================================================================
# EMA
# ============================================================================
def EMA(array, n):
    return list(pd.Series(array).ewm(span=n).mean())

# ============================================================================
# RSI
# ============================================================================
def RSI(array, n):
    # Approximate; good enough
    gain = pd.Series(array).diff()
    loss = gain.copy()
    gain[gain < 0] = 0
    loss[loss > 0] = 0
    rs = gain.ewm(n).mean() / loss.abs().ewm(n).mean()
    return 100 - 100 / (1 + rs)

# ============================================================================
# RCI:http://lowcost-greatidea.jp/technicalanalysis/rci/
# ============================================================================
def RCI(close, period):
    rank_period = np.arange(period, 0, -1)
    length = len(close)
    rci = np.zeros(length)
    
    for i in range(length):
        if i < period - 1:
            rci[i] = 0
        else :
            rank_price = pd.Series(close)[i - period + 1: i + 1].rank(method='min', ascending = False).values
            rci[i] = np.int32((1 - (6 * sum((rank_period - rank_price)**2)) / (period**3 - period)) * 100)
    return rci

# ============================================================================
# BB2
# ============================================================================
def bb2(array, n):
    gain=pd.DataFrame(array)
    gain.columns=['close']
    upper2, middle, lower2 = ta.BBANDS(gain.close, n,2,2,0)
    gain['bb_upper'] = upper2
    gain['bb_lower'] = lower2
    return gain['bb_upper'],gain['bb_lower']

# ============================================================================
# BB3
# ============================================================================
def bb3(array, n):
    gain=pd.DataFrame(array)
    gain.columns=['close']
    upper3, middle, lower3 = ta.BBANDS(gain.close, n,3,3,0)
    gain['bb_upper'] = upper3
    gain['bb_lower'] = lower3
    return gain['bb_upper'],gain['bb_lower']

## 1. Import data<a id='Import_data'></a>

In [2]:
inst = "USD_JPY"        # 米ドル
gran = "D"             # M30：30分、H1：1時間足、H4：4時間足、D：日足、W、週足、M、月足 
start_dt = "2010-07-26" # 再現性(SQN = 3.26達成)
end_dt = "2020-02-15"   # 再現性(原則、endは利用せず直近を取得し続ける)

oa = oandapy.API(environment="live", access_token=api_key)

def df_init(data):
    df = pd.DataFrame(data["candles"])
    df['time'] = df['time'].apply(lambda x: iso_to_jp(x))   # 日本時間に変換
    df['time'] = df['time'].apply(lambda x: date_to_str(x)) # 形式変換（文字列型）
    df["time"] = pd.to_datetime(df["time"])                  # 型変換
    df = df[["time","openAsk","highAsk","lowAsk","closeAsk"]].copy()
    df = df.rename(columns={"openAsk":"Open","highAsk":"High","lowAsk":"Low","closeAsk":"Close"})
    df = df.set_index("time")
    return df

data = oa.get_history(instrument=inst, granularity=gran, count=3000, start=start_dt)
df_test = df_init(data)
print(df_test.shape)
display(df_test.tail(6))

(2720, 4)


Open     High      Low    Close
time                                                   
2020-02-07 07:00:00  109.998  110.049  109.538  109.799
2020-02-10 07:00:00  109.705  109.885  109.569  109.783
2020-02-11 07:00:00  109.798  109.973  109.734  109.796
2020-02-12 07:00:00  109.793  110.181  109.783  110.144
2020-02-13 07:00:00  110.109  110.119  109.625  109.838
2020-02-14 07:00:00  109.839  109.918  109.705  109.796

## 2. Strategy<a id='Strategy'></a>

### 2-1. Best of RCIStrategy

In [173]:
class RCIStrategy(Strategy):

    n1 = 21; n2 = 75
    rci_s = 9; rci_m = 24; rci_l = 52
    rsi_m = 14;rsi_l = 42
    ema_s = 10
    
    def init(self):
        # SMA
        self.sma_s = self.I(SMA, self.data.Close, self.n1)
        self.sma_l = self.I(SMA, self.data.Close, self.n2)
        # RSI
        self.rsi_m = self.I(RSI, self.data.Close, self.rsi_m)
        self.rsi_l = self.I(RSI, self.data.Close, self.rsi_l)
        # BB
        self.bb_u2,self.bb_l2 = self.I(bb2,self.data.Close,  self.n1)
        self.bb_u3,self.bb_l3 = self.I(bb3,self.data.Close,  self.n1)
        # RCI
        self.rci_s = self.I(RCI, self.data.Close, self.rci_s)
        self.rci_m = self.I(RCI, self.data.Close, self.rci_m)
        self.rci_l = self.I(RCI, self.data.Close, self.rci_l)
        # EMA
        self.ema_s = self.I(EMA, self.data.Close, self.ema_s)
        
    def next(self):
        price = self.data.Close[-1]
        rci_s = self.rci_s[-1]
        rci_m = self.rci_m[-1]
        rci_l = self.rci_l[-1]
        rsi_m = self.rsi_m[-1]
        rsi_l = self.rsi_l[-1]
        ema_s = self.ema_s[-1]
        
        # *******************************************************************************
        # ■ ロング
        # *******************************************************************************
        if (self.position.is_short and ((1==0) 
            # -----------------------------------------------------------------------
            # ▼ Switch_Long ※ 保有中のショートを閉じてロングする
            # -----------------------------------------------------------------------
               or (rci_s < -87 and rci_m < 31 and rci_l < 73 and rci_l > -80) # Switch_Long1
               or (rci_s < -71 and rci_m > 75) # Switch_Long2
               or (rci_s < -70 and rci_l > 91)
             )):
            self.buy(sl=.98 * price)
        elif (not self.position and ((1==0) 
             # -----------------------------------------------------------------------
             # ▼ Entry Long ※理想は3重底
             # -----------------------------------------------------------------------
             or (rci_s < -90 and rci_m < -80) 
             or (rci_s < -72 and rci_m < 0 and rci_l > -71) 
             or (rci_s < -63 and rci_m < -35 and rci_m > -92 and rci_l > -75)
            )):
            self.buy(sl=.98 * price)
            
        elif (self.position.is_long and ((1==0) 
               # -----------------------------------------------------------------------
               # ▼ Exit Long ※理想は3天井
               # -----------------------------------------------------------------------
               or (rci_m > 88)
               or (rci_l > 75)
              )):
            self.position.close() 
            
        # *******************************************************************************
        # ■ ショート
        # *******************************************************************************
        if (self.position.is_long and ((1==0) 
               # -----------------------------------------------------------------------
               # ▼ Switch_Short ※ 保有中のロングを閉じてショートする
               # -----------------------------------------------------------------------
               or (rci_s > 80 and rci_m > 10)
               or (rci_s > 40 and rci_m < -80 and rci_l < -50) # 戻り売りチャンス
               or (rci_s > -90 and rci_m < -32 and crossover(self.sma_l, self.sma_s)) # デッドクロス
             )):
            self.sell(sl=1.02 * price)
        if (not self.position and ((1==0) 
             # -----------------------------------------------------------------------
             # ▼ Entry Short ※理想は3天井
             # -----------------------------------------------------------------------
             or (rci_s > 72 and rci_s < 95 and rci_m < 96 and rsi_l < 62) # ★★★
            )):
            self.sell(sl=1.02 * price)
            
        elif (self.position.is_short and ((1==0)
               # -----------------------------------------------------------------------
               # ▼ショートのExitルール(理想は3重底)
               # -----------------------------------------------------------------------
               or (rci_s < -91)
               or (rci_m < -86)
               or (rci_s < -70 and rci_m < -52 and rci_l < -30) 
              )):
            self.position.close()

### 2-2. Testing of RCIStrategy

In [188]:
class RCIStrategy2(Strategy):

    n1 = 21; n2 = 75
    rci_s = 9; rci_m = 24; rci_l = 52
    rsi_m = 14;rsi_l = 42
    ema_s = 10
    
    def init(self):
        # SMA
        self.sma_s = self.I(SMA, self.data.Close, self.n1)
        self.sma_l = self.I(SMA, self.data.Close, self.n2)
        # RSI
        self.rsi_m = self.I(RSI, self.data.Close, self.rsi_m)
        self.rsi_l = self.I(RSI, self.data.Close, self.rsi_l)
        # BB
        self.bb_u2,self.bb_l2 = self.I(bb2,self.data.Close,  self.n1)
        self.bb_u3,self.bb_l3 = self.I(bb3,self.data.Close,  self.n1)
        # RCI
        self.rci_s = self.I(RCI, self.data.Close, self.rci_s)
        self.rci_m = self.I(RCI, self.data.Close, self.rci_m)
        self.rci_l = self.I(RCI, self.data.Close, self.rci_l)
        # EMA
        self.ema_s = self.I(EMA, self.data.Close, self.ema_s)
        
    def next(self):
        price = self.data.Close[-1]
        rci_s = self.rci_s[-1]
        rci_m = self.rci_m[-1]
        rci_l = self.rci_l[-1]
        rsi_m = self.rsi_m[-1]
        rsi_l = self.rsi_l[-1]
        ema_s = self.ema_s[-1]
        
        # *******************************************************************************
        # ■ ロング
        # *******************************************************************************
        if (self.position.is_short and ((1==0) 
            # -----------------------------------------------------------------------
            # ▼ Switch_Long ※ 保有中のショートを閉じてロングする
            # -----------------------------------------------------------------------
               or (rci_s < -87 and rci_m < 31 and rci_l < 73 and rci_l > -80) # Switch_Long1
               or (rci_s < -71 and rci_m > 75) # Switch_Long2
               or (rci_s < -70 and rci_l > 91)
             )):
            self.buy(sl=.98 * price)
        elif (not self.position and ((1==0) 
             # -----------------------------------------------------------------------
             # ▼ Entry Long ※理想は3重底
             # -----------------------------------------------------------------------
             or (rci_s < -90 and rci_m < -80) 
             or (rci_s < -72 and rci_m < 0 and rci_l > -71) 
             or (rci_s < -63 and rci_m < -35 and rci_m > -92 and rci_l > -75)
            )):
            self.buy(sl=.98 * price)
            
        elif (self.position.is_long and ((1==0) 
               # -----------------------------------------------------------------------
               # ▼ Exit Long ※理想は3天井
               # -----------------------------------------------------------------------
               or (rci_m > 88)
               or (rci_l > 75)
              )):
            self.position.close() 
            
        # *******************************************************************************
        # ■ ショート
        # *******************************************************************************
        if (self.position.is_long and ((1==0) 
               # -----------------------------------------------------------------------
               # ▼ Switch_Short ※ 保有中のロングを閉じてショートする
               # -----------------------------------------------------------------------
               or (rci_s > 80 and rci_m > 10)
               or (rci_s > 40 and rci_m < -80 and rci_l < -50) # 戻り売りチャンス
               or (rci_s > -90 and rci_m < -32 and crossover(self.sma_l, self.sma_s)) # デッドクロス
             )):
            self.sell(sl=1.02 * price)
        if (not self.position and ((1==0) 
             # -----------------------------------------------------------------------
             # ▼ Entry Short ※理想は3天井
             # -----------------------------------------------------------------------
             or (rci_s > 72 and rci_s < 95 and rci_m < 96 and rsi_l < 62) # ★★★
            )):
            self.sell(sl=1.02 * price)
            
        elif (self.position.is_short and ((1==0)
               # -----------------------------------------------------------------------
               # ▼ショートのExitルール(理想は3重底)
               # -----------------------------------------------------------------------
               or (rci_s < -91)
               or (rci_m < -86)
               or (rci_s < -70 and rci_m < -52 and rci_l < -30) 
              )):
            self.position.close()

## 3. Backtest<a id='Backtest'></a>
#### ▼最高スコア ※2720
- Equity Final($) = 3450.26
- Trades = 147
- Win Rate(%) = 70.75
- Return(%) = 245.03
- SQN = 5.01
- Sharpe Ratio =  0.44
- Sortino Ratio =  1.14
- Calmar Ratio =  0.1
- Avg. Trade Duration = 19 days 00:00:00

In [195]:
bt = Backtest(df_test,
              RCIStrategy2,
              cash=1000,  # 所持金1000ドル(=約10万)
              commission=0.0002, # 取引手数料（為替価格に対する倍率で指定、為替価格100円でcommission=0.0002なら0.02円）
              trade_on_close=True # True：現在の終値に関してエントリー／ False：次の始値にエントリー
             )
out=bt.run()
print('- Equity Final($) =',round(out['Equity Final [$]'],2))
print('- Trades =',out['# Trades'])
print('- Win Rate(%) =',round(out['Win Rate [%]'],2))
print('- Return(%) =',round(out['Return [%]'],2))
print('- SQN =',round(out['SQN'],2))
print('- Sharpe Ratio = ',round(out['Sharpe Ratio'],2))
print('- Sortino Ratio = ',round(out['Sortino Ratio'],2))
print('- Calmar Ratio = ',round(out['Calmar Ratio'],2))
print('- Avg. Trade Duration =',out['Avg. Trade Duration'])
bt.plot()

- Equity Final($) = 3450.26
- Trades = 147
- Win Rate(%) = 70.75
- Return(%) = 245.03
- SQN = 5.01
- Sharpe Ratio =  0.44
- Sortino Ratio =  1.14
- Calmar Ratio =  0.1
- Avg. Trade Duration = 19 days 00:00:00


In [104]:
print(out)

Start                     2010-07-26 06:00:00
End                       2020-02-14 07:00:00
Duration                   3490 days 01:00:00
Exposure [%]                          84.3567
Equity Final [$]                      2796.45
Equity Peak [$]                       2802.81
Return [%]                            179.645
Buy & Hold Return [%]                 26.3664
Max. Drawdown [%]                    -11.3043
Avg. Drawdown [%]                    -1.14421
Max. Drawdown Duration      323 days 00:00:00
Avg. Drawdown Duration       25 days 00:00:00
# Trades                                  166
Win Rate [%]                          65.6627
Best Trade [%]                        6.61998
Worst Trade [%]                       -2.0196
Avg. Trade [%]                       0.640836
Max. Trade Duration          95 days 00:00:00
Avg. Trade Duration          18 days 00:00:00
Expectancy [%]                        1.69726
SQN                                   4.15973
Sharpe Ratio                      

|項目|項目の見方|
| :- | :- | 
|**Start**|ヒストリカルデータ開始日時|
|**End**|ヒストリカルデータ終了日時|
|**Duration**|ヒストリカルデータ期間|
|**Exposure (%)** |ポジション保有期間の割合|
|**Equity Final**|最終金額|
|**Equity Peak**|最高金額|
|**Return (%)**|利益率=損益÷開始時所持金×100|
|**Buy & Hold Return (%)**|((終了時の終値 開始時の終値)÷開始時の終値)の絶対値×100|
|**Max. Drawdown (%)**|最大下落率|
|**Avg. Drawdown (%)**|平均下落率|
|**Max. Drawdown Duration**|最大下落期間|
|**Avg. Drawdown Duration**|平均下落期間|
|**Trades**| 取引回数|
|**Win Rate (%)**|勝率=勝ち取引回数÷全取引回数×100|
|**Best Trade (%)**|1回の取引での利益の最大値÷所持金×100|
|**Worst Trade (%)**|1回の取引での損失の最大値÷所持金×100|
|**Avg. Trade (%)**|損益の平均値÷所持金×100|
|**Max. Trade Duration**|1回の取引での最長期間|
|**Avg. Trade Duration**|1回の取引での平均期間|
|**Expectancy (%)**|期待値=平均利益×勝率＋平均損失×敗率（１取引で期待できる利益、正は資産が増え、負は資産が減る。）|
|**SQN**|「期待値/標準偏差×√100」で計算されるシステムの品質の評価値。評価方法は、1.6～平均以下、2.0～平均、2.5～良い、3.0～優秀、5.1～最高、7.0～聖杯　※取引数が30以上で、SQN値は信頼できる。|
|**Sharpe Ratio**|シャープレシオとは利益とリスクの比率のことで、値が大きいほど資産曲線がなめらかになり安定性のある利益が見込めます。|
|**Sortino Ratio**|シャープレシオだけでは分からない下方リスクの抑制度合い。通常、この数値が大きいほど優れている（下落局面に強い）|
|**Calmar Ratio**|最大損失率に対する年間平均収益の比率 ※値が低いほど指定された期間に渡ってリスク調整ベースで実行された投資は悪化し、値が高いほどパフォーマンスが向上する。|

## 4. Optimisation<a id='optimisation'></a>

In [7]:
out2=bt.optimize(rci_s=range(9, 10),rci_m=range(24, 25),rci_l=range(52, 53))

In [82]:
print('▼最適化結果 ※',out2._trade_data.shape)
print('Equity Final($)=',round(out2['Equity Final [$]'],2))
print('Trades=',out2['# Trades'])
print('Win Rate(%)=',round(out2['Win Rate [%]'],2))
print('Return(%)=',round(out2['Return [%]'],2))
print('SQN=',round(out2['SQN'],2))
print('_strategy=',out2['_strategy'])

▼最適化結果 ※ (2720, 9)
Equity Final($)= 2309.15
Trades= 186
Win Rate(%)= 63.44
Return(%)= 130.91
SQN= 3.26
_strategy= RCIStrategy2(rci_s=9,rci_m=24,rci_l=52)


## 5. Output Excel Data<a id='Output'><a>

In [194]:
pd.set_option('display.max_rows', 100)

xls_df = df_test
xls_df['RCI_s'] = RCI(xls_df.Close,9)
xls_df['RCI_m'] = RCI(xls_df.Close,24)
xls_df['RCI_l'] = RCI(xls_df.Close,52)
xls_df['RSI_m'] = RSI(xls_df.Close,14)
xls_df['RSI_l'] = RSI(xls_df.Close,42)
xls_df['SMA_s'] = SMA(xls_df.Close,21)
xls_df['SMA_l'] = SMA(xls_df.Close,75)
xls_df['EMA_s'] = EMA(xls_df.Close,10)

df_concat = pd.concat([out._trade_data, xls_df], axis=1)
df_concat = df_concat.fillna({'SMA_s': 0.0})
df_concat = df_concat.fillna({'SMA_l': 0.0})
df_concat = df_concat[['Entry Price','Exit Price','Returns','Open','High','Low','Close','RCI_s','RCI_m','RCI_l','RSI_m','RSI_l','SMA_s','SMA_l','EMA_s']]

df_concat.to_excel("./backtest_" + str(inst) + "_" + str(gran) + ".xlsx",startrow=0, startcol=0)
display(df_concat.tail(6))

Entry Price  Exit Price  Returns     Open     High  \
2020-02-07 07:00:00          NaN         NaN      NaN  109.998  110.049   
2020-02-10 07:00:00          NaN         NaN      NaN  109.705  109.885   
2020-02-11 07:00:00          NaN         NaN      NaN  109.798  109.973   
2020-02-12 07:00:00          NaN         NaN      NaN  109.793  110.181   
2020-02-13 07:00:00          NaN         NaN      NaN  110.109  110.119   
2020-02-14 07:00:00          NaN         NaN      NaN  109.839  109.918   

                         Low    Close  RCI_s  RCI_m  RCI_l      RSI_m  \
2020-02-07 07:00:00  109.538  109.799   63.0  -16.0   27.0  55.679852   
2020-02-10 07:00:00  109.569  109.783   70.0  -27.0   30.0  55.439984   
2020-02-11 07:00:00  109.734  109.796   65.0  -33.0   35.0  55.606471   
2020-02-12 07:00:00  109.783  110.144   70.0  -26.0   41.0  59.903256   
2020-02-13 07:00:00  109.625  109.838   50.0  -33.0   45.0  54.897371   
2020-02-14 07:00:00  109.705  109.796   19.0  -27.0   46.0  54.230920   

                         RSI_l       SMA_s       SMA_l       EMA_s  
2020-02-07 07:00:00  53.852778  109.566952  109.132867  109.485590  
2020-02-10 07:00:00  53.768544  109.580619  109.148147  109.539664  
2020-02-11 07:00:00  53.828619  109.573190  109.162947  109.586271  
2020-02-12 07:00:00  55.416389  109.579571  109.178480  109.687676  
2020-02-13 07:00:00  53.752312  109.576095  109.190987  109.715008  
2020-02-14 07:00:00  53.526447  109.557571  109.203240  109.729734

## 6. Appendix. Macro to attach to exported Excel<a id='Vba'><a>

Option Explicit

    Private sh As Worksheet
    Private entry_dic, exit_dic, win_entry_dic, lse_entry_dic, win_exit_dic, lse_exit_dic As Object
    Private row_min, row_max, col_max As Integer
    
Private Enum COL_
    '---------------------------
    Time = 1     'A
    entry_p1     'B
    exit_p       'C
    Rtns         '
    '---------------------------
    Opn          '
    Hih          '
    Low          '
    Cls          '
    '---------------------------
    rci_s        '
    rci_m        '
    rci_l        '
    rsi_m
    rsi_l
    sma_s        '
    sma_l        '
    ema_s        '
    '---------------------------
    entry_nm1    'O
    position     'P
    win_pips2
    pos_pips
    trade_no
    result       '
    pos_days
    entry_nm2
    win_pips1
    exit_nm
End Enum

'http://excel.style-mods.net/topic11.htm
Private Enum FClr_
    Exi = 3
    Loss = 26
End Enum
Private Enum BClr_
    Header = 24      'ヘッダ
    Entry = 38       'エントリー条件
    win_long = 34    '勝ちロング
    lse_long = 28    '負けロング
    win_short = 19   '勝ちショート
    lse_short = 27   '負けショート
End Enum

Public Sub CreateTradeResult()
    Application.ScreenUpdating = False

    Set sh = ActiveSheet
    
    sh.Cells.ClearFormats
    sh.Rows("1:1").Font.Bold = True
    sh.Range("A1").Value = "Time"
    
    ' ----------------------------------------------
    sh.Columns(GetAlp(COL_.entry_nm1) + ":BZ").ClearContents
    
    sh.Cells(1, COL_.entry_nm1).Value = "Entry"
    sh.Cells(1, COL_.position).Value = "Position"
    sh.Cells(1, COL_.win_pips2).Value = "Res_pips"
    sh.Cells(1, COL_.pos_pips).Value = "Pos_pips"
    sh.Cells(1, COL_.pos_days).Value = "Days"
    sh.Cells(1, COL_.trade_no).Value = "Trade No"
    
    sh.Cells(1, COL_.result).Value = "Result"
    sh.Cells(1, COL_.entry_nm2).Value = "Entry(last)"
    sh.Cells(1, COL_.win_pips1).Value = "Get_pips"
    sh.Cells(1, COL_.exit_nm).Value = "Exit"
    
    sh.Columns("A:A").NumberFormatLocal = "yy-mm-dd aaa hh"
    sh.Columns("B:H").NumberFormatLocal = "0.000"
    sh.Columns("L:P").NumberFormatLocal = "0.000"
    
    Dim i, clr_no, trade_no, win_pips As Integer
    Dim last_position, last_trade_no, last_entry_p, last_entry_name, last_entry_row, result As String
    Dim position, s_trade_no, entry_p, exit_p, rtn_val, entry_name, exit_name, close_p As String
    Dim last_pos_long, last_pos_short As Boolean
    
    Dim rci_s, rci_m, rci_l As Integer
    Dim rsi_m, rsi_l As Single
    Dim sma_s, sma_l As Single
    
    Call entry_dic_init
    Call exit_dic_init
    
    row_min = 2
    row_max = sh.Cells(1, 1).End(xlDown).row
    col_max = sh.Cells(1, 1).End(xlToRight).Column
    
    last_entry_p = "0"
    last_entry_name = ""
    last_entry_row = "0"
    last_position = ""
    last_trade_no = 0
    trade_no = 0

    For i = row_min To row_max

        entry_name = ""
        exit_name = ""
        position = ""
        s_trade_no = ""
        
        win_pips = 0
        
        entry_p = sh.Cells(i, COL_.entry_p1).Value
        exit_p = sh.Cells(i, COL_.exit_p).Value
        rtn_val = sh.Cells(i, COL_.Rtns).Value
        close_p = sh.Cells(i, COL_.Cls).Value
        rci_s = CInt(sh.Cells(i, COL_.rci_s).Value)
        rci_m = CInt(sh.Cells(i, COL_.rci_m).Value)
        rci_l = CInt(sh.Cells(i, COL_.rci_l).Value)
        rsi_m = CSng(sh.Cells(i, COL_.rsi_m).Value)
        rsi_l = CSng(sh.Cells(i, COL_.rsi_l).Value)
        sma_s = CSng(sh.Cells(i, COL_.sma_s).Value)
        sma_l = CSng(sh.Cells(i, COL_.sma_l).Value)
        
        last_pos_long = InStr(UCase(last_position), "LONG") > 0
        last_pos_short = InStr(UCase(last_position), "SHORT") > 0
        
        ' ※行の種類
        ' 1.エントリーのみ ---------------> 判断可能
        ' 2.決済のみ       ---------------> 判断可能
        ' 3.エントリーと決済       -------> 判断可能
        ' 4.保有中     -------------------> エントリー時にポジションを記載することで判断
        ' 5.保有なし   -------------------> 上記以外
        
        '********* debug用 *********
        If last_trade_no = 764 Then
            Debug.Print "▼" + Str(i) + "行目 T" + CStr(last_trade_no)
        End If
        '***************************
        
        ' ===============================================
        ' ■ エントリー情報表示
        ' ===============================================
        If (entry_p <> "" And rtn_val = "") Or (entry_p <> "" And rtn_val <> "") Then
            ' ▼ 1.エントリーのみ + ▼ 3.エントリーと決済
            trade_no = trade_no + 1
            s_trade_no = CStr(trade_no)
            entry_name = GetEntryName(i, rci_s, rci_m, rci_l, last_position, sma_s, sma_l, rsi_m, rsi_l)
            position = entry_name
            
            sh.Cells(i, COL_.entry_nm1).Value = entry_name
            sh.Cells(i, COL_.position).Value = position
            sh.Cells(i, COL_.trade_no).Value = IIf(last_trade_no <> 0, "T" + CStr(last_trade_no) + ",T" + CStr(trade_no), "T" + CStr(trade_no))
            
        ElseIf entry_p = "" And rtn_val <> "" Then
            ' ▼ 2.決済のみ
            sh.Cells(i, COL_.entry_nm1).Value = entry_name
            sh.Cells(i, COL_.position).Value = last_position
            sh.Cells(i, COL_.trade_no).Value = "T" + CStr(last_trade_no)
            sh.Cells(i, COL_.pos_pips).Value = IIf(last_pos_long, CInt((CSng(close_p) - CSng(last_entry_p)) * 100), CInt((CSng(last_entry_p) - CSng(close_p)) * 100))
            
        ElseIf last_pos_long Or last_pos_short Then
            ' ▼ 4.保有中
            sh.Cells(i, COL_.entry_nm1).Value = entry_name
            sh.Cells(i, COL_.position).Value = last_position
            sh.Cells(i, COL_.trade_no).Value = "T" + CStr(last_trade_no)
            sh.Cells(i, COL_.pos_pips).Value = IIf(last_pos_long, CInt((CSng(close_p) - CSng(last_entry_p)) * 100), CInt((CSng(last_entry_p) - CSng(close_p)) * 100))
            
        Else
            ' ▼ 5.保有なし
            sh.Cells(i, COL_.entry_nm1).Value = "-"
            sh.Cells(i, COL_.position).Value = "-"
            sh.Cells(i, COL_.trade_no).Value = "-"
        End If
                
        ' ===============================================
        ' ■ 決済情報表示
        ' ===============================================
        If rtn_val <> "" Then
            If CSng(rtn_val) > 0 Then
                ' 勝ち
                win_pips = IIf(last_pos_long, Int((exit_p - last_entry_p) * 100), Int((last_entry_p - exit_p) * 100))
                ' entry_nameに値があれば、同時エントリー（前回Exitがswitchかを判断）
                exit_name = GetExitName(i, last_pos_long, win_pips, entry_name, rci_s, rci_m, rci_l)
            
                win_entry_dic(last_entry_name) = win_entry_dic(last_entry_name) + 1
                win_exit_dic(exit_name) = win_exit_dic(exit_name) + 1
                
            Else
                ' 負け
                win_pips = IIf(last_pos_long, Int((exit_p - last_entry_p) * 100), Int((last_entry_p - exit_p) * 100))
                ' entry_nameに値があれば、同時エントリー（前回Exitがswitchかを判断）
                exit_name = GetExitName(i, last_pos_long, win_pips, entry_name, rci_s, rci_m, rci_l)
                lse_entry_dic(last_entry_name) = lse_entry_dic(last_entry_name) + 1
                lse_exit_dic(exit_name) = lse_exit_dic(exit_name) + 1
                
            End If ' rtn_val > 0

            sh.Cells(i, COL_.result).Value = GetResultChar(rtn_val, last_pos_long)
            sh.Cells(i, COL_.entry_nm2).Value = last_entry_name
            sh.Cells(i, COL_.win_pips1).Value = win_pips
            sh.Cells(i, COL_.exit_nm).Value = exit_name
            sh.Cells(i, COL_.pos_days).Value = (i - last_entry_row) * 4 / 24  '保有日数
            sh.Cells(i, COL_.pos_days).NumberFormatLocal = "0.00"
            
            sh.Cells(last_entry_row, COL_.win_pips2).Value = win_pips
            
            ' 書式設定
            sh.Range(Cells(last_entry_row, COL_.entry_nm1), Cells(i, COL_.trade_no)).Interior.ColorIndex = GetResultColor(rtn_val, last_pos_long) 'エントリー行
            sh.Range(Cells(i, COL_.result), Cells(i, COL_.exit_nm)).Interior.ColorIndex = GetResultColor(rtn_val, last_pos_long)         '決済行
            
            If InStr(exit_name, "Loss") > 0 Then
                'ロスカット文字色変更
                sh.Range(Cells(last_entry_row, COL_.entry_nm1), Cells(i, COL_.trade_no)).Font.ColorIndex = FClr_.Loss  'エントリー行
                sh.Range(Cells(i, COL_.result), Cells(i, COL_.exit_nm)).Font.ColorIndex = FClr_.Loss      '決済行
            End If
            
            
        End If ' rtn_val <> ""
        
        ' 過去データ更新
        If entry_p <> "" Then
            ' ▼ エントリーのみ + ▼エントリーと決済
            last_position = position
            last_entry_p = entry_p
            last_entry_name = entry_name
            last_entry_row = i
            last_trade_no = s_trade_no ' 前回トレード番号更新
            
        ElseIf entry_p = "" And rtn_val <> "" Then
            ' ▼決済のみ
            last_position = ""
            last_entry_p = 0
            last_entry_name = ""
            last_entry_row = -1
            last_trade_no = 0
        End If
        
    Next i
    
    ' 集計
    Call ShowAggregate
    
    ' 体裁
    sh.Range(Cells(1, 1), Cells(row_max, col_max)).Borders.LineStyle = True
    sh.Range(Cells(1, 1), Cells(row_max, col_max)).Font.Size = 10
    sh.Range(Cells(1, 1), Cells(row_max, col_max)).Font.Name = "ＭＳ ゴシック"
    sh.Range(Cells(1, 1), Cells(1, col_max)).Interior.ColorIndex = BClr_.Header
    
    ' 列幅設定
    Cells.EntireColumn.AutoFit
    
    'ヘッダ固定
    Range("B2").Select
    ActiveWindow.FreezePanes = True
    
    Set entry_dic = Nothing
    Set exit_dic = Nothing
    Set win_entry_dic = Nothing
    Set lse_entry_dic = Nothing
    Set win_exit_dic = Nothing
    Set lse_exit_dic = Nothing
    
    Application.Goto Reference:=Range(GetAlp(COL_.entry_nm1) + CStr(row_max + 2)), Scroll:=True
    Application.ScreenUpdating = False
        
End Sub

Private Function GetResultChar(rtn_val, last_pos_long) As String
    Dim rtn As String
    rtn = ""
    If CSng(rtn_val) > 0 Then
        If last_pos_long Then
            rtn = "○win_long"
        Else
            rtn = "○win_short"
        End If
    Else
        If last_pos_long Then
            rtn = "●lose_long"
        Else
            rtn = "●lose_short"
        End If
    End If
    
    GetResultChar = rtn
    
End Function

Private Function GetResultColor(rtn_val, last_pos_long) As Integer
    Dim rtn As Integer
    rtn = 0
    If CSng(rtn_val) > 0 Then
        If last_pos_long Then
            rtn = BClr_.win_long
        Else
            rtn = BClr_.win_short
        End If
    Else
        If last_pos_long Then
            rtn = BClr_.lse_long
        Else
            rtn = BClr_.lse_short
        End If
    End If
    GetResultColor = rtn
    
End Function


'# *******************************************************************************
'# ■ エントリー条件
'# *******************************************************************************
Private Sub entry_dic_init()
    Set entry_dic = CreateObject("Scripting.Dictionary")
    Set win_entry_dic = CreateObject("Scripting.Dictionary")
    Set lse_entry_dic = CreateObject("Scripting.Dictionary")
    
    ' ↓※Switchの場合は、エクジット条件のほうも修正する必要があるので注意
    ' ========================================================================================
    ' ▼ ●●●● ▼ エントリーロング ▼ ●●●● ▼
    entry_dic.Add "Entry_Long1", "通常ロング1"
    entry_dic.Add "Entry_Long2", "通常ロング2"
    entry_dic.Add "Entry_Long3", "通常ロング3"
    ' Switch
    entry_dic.Add "Switch_Long1", "切替ロング1"
    entry_dic.Add "Switch_Long2", "切替ロング2"
    entry_dic.Add "Switch_Long3", "切替ロング3"
    
    ' ▼ ■■■■ ▼ エントリーショート ▼ ■■■■ ▼
    entry_dic.Add "Entry_Short1", "通常ショート1"
    ' Switch
    entry_dic.Add "Switch_Short1", "切替ショート1"
    entry_dic.Add "Switch_Short2", "切替ショート2"
    entry_dic.Add "Switch_Short3", "切替ショート3"
    ' ========================================================================================
    
    Dim curKey As Variant
    For Each curKey In entry_dic
        win_entry_dic.Add curKey, 0 '初期化
        lse_entry_dic.Add curKey, 0 '初期化
    Next
    
End Sub

Private Function GetEntryName(row, rci_s, rci_m, rci_l, position, sma_s, sma_l, rsi_m, rsi_l) As String
    Dim rtn As String
    Dim pos_is_short, pos_is_long As Boolean
    rtn = ""
    pos_is_long = InStr(UCase(position), "LONG") > 0
    pos_is_short = InStr(UCase(position), "SHORT") > 0
    
    ' ========================================================================================
    If 1 = 0 Then
    ' ▼ ●●●● ▼ エントリーロング ▼ ●●●● ▼
    ' Switch ※ 決済直後の通常エントリーの場合、前回保有情報が消えていないため、先にSwitchロジックを記載する。
    ElseIf pos_is_short And rci_s < -87 And rci_m < 31 And rci_l < 73 And rci_l > -80 Then: rtn = "Switch_Long1" '< ------------- Switch_Long1
        sh.Cells(row, COL_.rci_s).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_m).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_l).Interior.ColorIndex = BClr_.Entry
    
    ElseIf pos_is_short And rci_s < -71 And rci_m > 75 Then: rtn = "Switch_Long2" '< --------------------------- Switch_Long2
        sh.Cells(row, COL_.rci_s).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_m).Interior.ColorIndex = BClr_.Entry
    
    ElseIf pos_is_short And rci_s < -70 And rci_l > 91 Then: rtn = "Switch_Long3" '< --------------------------- Switch_Long3
        sh.Cells(row, COL_.rci_s).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_l).Interior.ColorIndex = BClr_.Entry
    
        
    ' 通常
    ElseIf rci_s < -90 And rci_m < -80 Then: rtn = "Entry_Long1" '< ----------------------------------------- Entry_Long1
        sh.Cells(row, COL_.rci_s).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_m).Interior.ColorIndex = BClr_.Entry
        
    ElseIf rci_s < -72 And rci_m < 0 And rci_l > -71 Then: rtn = "Entry_Long2"  '< --------------------------- Entry_Long2
        sh.Cells(row, COL_.rci_s).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_m).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_l).Interior.ColorIndex = BClr_.Entry
        
    ElseIf rci_s < -63 And rci_m < -35 And rci_m > -92 And rci_l > -75 Then: rtn = "Entry_Long3" '< --------------------------- Entry_Long3
        sh.Cells(row, COL_.rci_s).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_m).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_l).Interior.ColorIndex = BClr_.Entry
    End If
    
    If 1 = 0 Then
    ' ▼ ■■■■ ▼ エントリーショート ▼ ■■■■ ▼
    ' Switch ※ 決済直後の通常エントリーの場合、前回保有情報が消えていないため、先にSwitchロジックを記載する。
    ElseIf pos_is_long And rci_s > 80 And rci_m > 10 Then: rtn = "Switch_Short1" '< ---------- Switch_Short1
        sh.Cells(row, COL_.rci_s).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_m).Interior.ColorIndex = BClr_.Entry
        
    ElseIf pos_is_long And rci_s > 40 And rci_m < -80 And rci_l < -50 Then: rtn = "Switch_Short2" '< ------ Switch_Short2
        sh.Cells(row, COL_.rci_s).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_m).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_l).Interior.ColorIndex = BClr_.Entry
        
    ElseIf pos_is_long And rci_s > -90 And rci_m < -32 And sma_s < sma_l Then: rtn = "Switch_Short3" '< ------ Switch_Short3
        sh.Cells(row, COL_.rci_s).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_m).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.sma_s).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.sma_l).Interior.ColorIndex = BClr_.Entry
        
    ' 通常
    ElseIf rci_s > 72 And rci_s < 95 And rci_m < 96 And rsi_l < 62 Then: rtn = "Entry_Short1"    '< ------------ Entry_Short1
        sh.Cells(row, COL_.rci_s).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rci_m).Interior.ColorIndex = BClr_.Entry
        sh.Cells(row, COL_.rsi_l).Interior.ColorIndex = BClr_.Entry
        
    End If
    ' ========================================================================================
    
    If rtn = "" Then
        Debug.Print "Entry Logic Not Found.", row, rci_s, rci_m, rci_l, rsi_l
    End If
    
    GetEntryName = rtn
    
End Function

'# *******************************************************************************
'# ■ エクジット条件
'# *******************************************************************************
Private Sub exit_dic_init()
    Set exit_dic = CreateObject("Scripting.Dictionary")
    Set lse_exit_dic = CreateObject("Scripting.Dictionary")
    Set win_exit_dic = CreateObject("Scripting.Dictionary")

    ' ↓※Switchの場合は、エントリー条件のほうも修正する必要があるので注意
    ' ========================================================================================
    ' ▼ ◆◆◆◆ ▼ エクジットロング ▼ ◆◆◆◆ ▼
    exit_dic.Add "Exit_Long1", "通常エクジットロング1"
    exit_dic.Add "Exit_Long2", "通常エクジットロング2"
    ' Switch
    exit_dic.Add "Switch_Short1", "ショート切替1"
    exit_dic.Add "Switch_Short2", "ショート切替2"
    exit_dic.Add "Switch_Short3", "ショート切替3"
    ' Loss_cut
    exit_dic.Add "★Loss_cut_Long", "ロスカット強制終了"
    
    ' ▼ ★★★★ ▼ エクジットショート ▼ ★★★★ ▼
    exit_dic.Add "Exit_Short1", "通常エクジットショート1"
    exit_dic.Add "Exit_Short2", "通常エクジットショート2"
    exit_dic.Add "Exit_Short3", "通常エクジットショート3"
    ' Switch
    exit_dic.Add "Switch_Long1", "ロング切替"
    exit_dic.Add "Switch_Long2", "ロング切替"
    exit_dic.Add "Switch_Long3", "ロング切替"
    ' Loss_cut
    exit_dic.Add "★Loss_cut_Short", "ロスカット強制終了"
    ' ========================================================================================
    
    Dim curKey As Variant
    For Each curKey In exit_dic
        win_exit_dic.Add curKey, 0 '初期化
        lse_exit_dic.Add curKey, 0 '初期化
    Next
    
End Sub

Private Function GetExitName(row, long_flg, win_pips, entry_name, rci_s, rci_m, rci_l) As String
    Dim rtn As String
    Dim switch_flg, re_entry_flg As Boolean
    rtn = ""
    
    switch_flg = InStr(entry_name, "Switch") > 0
    re_entry_flg = InStr(entry_name, "ReEntry") > 0
    
    ' ========================================================================================
    If long_flg Then
        If 1 = 0 Then
        ' ▼ ◆◆◆◆ ▼ エクジットロング ▼ ◆◆◆◆ ▼
        ElseIf switch_flg Then: rtn = entry_name
        ElseIf re_entry_flg Then: rtn = entry_name
        '通常
        ElseIf rci_m > 88 Then: rtn = "Exit_Long1"  '< --------------------------- Exit_Long1
            sh.Cells(row, COL_.rci_m).Font.ColorIndex = FClr_.Exi
            
        ElseIf rci_l > 75 Then: rtn = "Exit_Long2"  '< --------------------------- Exit_Long2
            sh.Cells(row, COL_.rci_l).Font.ColorIndex = FClr_.Exi
        
        'ロスカット
        ElseIf win_pips < -100 Then: rtn = "★Loss_cut_Long" '< ------------------ ★Loss_cut_Long
        Else:
        End If
    
    Else
        If 1 = 0 Then
        ' ▼ ★★★★ ▼ エクジットショート ▼ ★★★★ ▼
        ElseIf switch_flg Then: rtn = entry_name
        ElseIf re_entry_flg Then: rtn = entry_name
        '通常
        ElseIf rci_s < -91 Then: rtn = "Exit_Short1" '< --------------------------- Exit_Long1
            sh.Cells(row, COL_.rci_s).Font.ColorIndex = FClr_.Exi

        ElseIf rci_m < -86 Then: rtn = "Exit_Short2"      '< ---------------------- Exit_Short2
            sh.Cells(row, COL_.rci_m).Font.ColorIndex = FClr_.Exi
        
        ElseIf rci_s < -70 And rci_m < -52 And rci_l < -30 Then: rtn = "Exit_Short3" '< -------------------- Exit_Short3
            sh.Cells(row, COL_.rci_s).Font.ColorIndex = FClr_.Exi
            sh.Cells(row, COL_.rci_m).Font.ColorIndex = FClr_.Exi
            sh.Cells(row, COL_.rci_l).Font.ColorIndex = FClr_.Exi
        
        'ロスカット
        ElseIf win_pips < -100 Then: rtn = "★Loss_cut_Short" '< ---------------------- ★Loss_cut_Short
        Else:
        End If
    End If
    ' ========================================================================================
                    
    If rtn = "" Then
        rtn = "★ Error : Not Found Exit Name ★"
    End If
    
    GetExitName = rtn
    
End Function

Private Sub ShowAggregate()

    Dim alp1, alp2, alp3, alp4 As String
    Dim curKey As Variant
    Dim row_idx, k, win_cnt, lose_cnt As Integer
    Dim s_cell As Range
    
    ' ==============================================================
    ' Entry Logic
    ' ==============================================================
    Set s_cell = sh.Range(GetAlp(COL_.entry_nm1) + CStr(row_max + 3))
    row_idx = 1
    s_cell.Offset(1, 0).Value = "▼ Entry Logic": s_cell.Offset(1, 1).Value = "回数"
    s_cell.Offset(1, 2).Value = "勝": s_cell.Offset(1, 3).Value = "負": s_cell.Offset(1, 4).Value = "勝率"
    s_cell.Offset(1, 5).Value = "ave.pips": s_cell.Offset(1, 6).Value = "ave.days"
    
    alp1 = GetAlp(COL_.entry_nm1) 'エントリー列1 O
    alp2 = GetAlp(COL_.win_pips2) '獲得pips　Q
    alp3 = GetAlp(COL_.entry_nm2) 'エントリー列2 V
    alp4 = GetAlp(COL_.pos_days)  '保有日数 U
    
    For Each curKey In entry_dic
        row_idx = row_idx + 1
        s_cell.Offset(row_idx, 0).Value = curKey
        s_cell.Offset(row_idx, 1).Formula = "=COUNTIF(" + alp1 + "2:" + alp1 + CStr(row_max) + ", " + CStr(s_cell.Offset(row_idx, 0).Address) + ")"
        s_cell.Offset(row_idx, 5).Formula = "=SUMIF(" + alp1 + "2:" + alp1 + CStr(row_max) + "," + CStr(s_cell.Offset(row_idx, 0).Address) + "," + alp2 + "2:" + alp2 + CStr(row_max) + ")/" + CStr(s_cell.Offset(row_idx, 1).Address)
        s_cell.Offset(row_idx, 6).Formula = "=SUMIF(" + alp3 + "2:" + alp3 + CStr(row_max) + "," + CStr(s_cell.Offset(row_idx, 0).Address) + "," + alp4 + "2:" + alp4 + CStr(row_max) + ")/" + CStr(s_cell.Offset(row_idx, 1).Address)
    Next
    For Each curKey In win_entry_dic
        For k = 2 To entry_dic.Count + 1
            If curKey = s_cell.Offset(k, 0).Value Then
                win_cnt = CInt(win_entry_dic.Item(curKey))
                lose_cnt = CInt(lse_entry_dic.Item(curKey))
                s_cell.Offset(k, 2).Value = win_cnt
                s_cell.Offset(k, 3).Value = lose_cnt
                If Not (win_cnt = 0 And lose_cnt = 0) Then
                    s_cell.Offset(k, 4).Value = win_cnt / (win_cnt + lose_cnt) * 100
                End If
            End If
        Next k
    Next
    
    '合計
    s_cell.Offset(row_idx + 1, 0).Value = "合計"
    s_cell.Offset(row_idx + 1, 1).Formula = "=SUM(" + CStr(s_cell.Offset(2, 1).Address) + ":" + CStr(s_cell.Offset(row_idx, 1).Address) + ")"
    s_cell.Offset(row_idx + 1, 2).Formula = "=SUM(" + CStr(s_cell.Offset(2, 2).Address) + ":" + CStr(s_cell.Offset(row_idx, 2).Address) + ")"
    s_cell.Offset(row_idx + 1, 3).Formula = "=SUM(" + CStr(s_cell.Offset(2, 3).Address) + ":" + CStr(s_cell.Offset(row_idx, 3).Address) + ")"
    
    sh.Range(s_cell.Offset(1, 0), s_cell.Offset(1, 6)).Interior.ColorIndex = BClr_.Header
    sh.Range(s_cell.Offset(1, 0), s_cell.Offset(row_idx + 1, 6)).Borders.LineStyle = True
    sh.Range(s_cell.Offset(1, 4), s_cell.Offset(row_idx + 1, 6)).NumberFormatLocal = "0.0"
    
    ' ==============================================================
    ' Exit Logic
    ' ==============================================================
    Set s_cell = sh.Range(GetAlp(COL_.entry_nm1) + CStr(row_max + 20))
    row_idx = 1
    s_cell.Offset(1, 0).Value = "▼ Exit Logic": s_cell.Offset(1, 1).Value = "回数"
    s_cell.Offset(1, 2).Value = "勝": s_cell.Offset(1, 3).Value = "負": s_cell.Offset(1, 4).Value = "勝率"
    s_cell.Offset(1, 5).Value = "ave.pips": s_cell.Offset(1, 6).Value = "ave.days"
    
    alp1 = GetAlp(COL_.exit_nm)
    alp2 = GetAlp(COL_.win_pips1) '獲得pips
    alp3 = GetAlp(COL_.exit_nm) 'エントリー列2
    alp4 = GetAlp(COL_.pos_days)  '保有日数

    For Each curKey In exit_dic
        row_idx = row_idx + 1
        s_cell.Offset(row_idx, 0).Value = curKey
        s_cell.Offset(row_idx, 1).Formula = "=COUNTIF(" + alp1 + "2:" + alp1 + CStr(row_max) + ", " + CStr(s_cell.Offset(row_idx, 0).Address) + ")"
        s_cell.Offset(row_idx, 5).Formula = "=SUMIF(" + alp1 + "2:" + alp1 + CStr(row_max) + "," + CStr(s_cell.Offset(row_idx, 0).Address) + "," + alp2 + "2:" + alp2 + CStr(row_max) + ")/" + CStr(s_cell.Offset(row_idx, 1).Address)
        s_cell.Offset(row_idx, 6).Formula = "=SUMIF(" + alp3 + "2:" + alp3 + CStr(row_max) + "," + CStr(s_cell.Offset(row_idx, 0).Address) + "," + alp4 + "2:" + alp4 + CStr(row_max) + ")/" + CStr(s_cell.Offset(row_idx, 1).Address)
    Next
    For Each curKey In win_exit_dic
        For k = 2 To exit_dic.Count + 1
            If curKey = s_cell.Offset(k, 0).Value Then
                win_cnt = CInt(win_exit_dic.Item(curKey))
                lose_cnt = CInt(lse_exit_dic.Item(curKey))
                s_cell.Offset(k, 2).Value = win_cnt
                s_cell.Offset(k, 3).Value = lose_cnt
                If Not (win_cnt = 0 And lose_cnt = 0) Then
                    s_cell.Offset(k, 4).Value = win_cnt / (win_cnt + lose_cnt) * 100
                End If
            End If
        Next k
    Next
    
    '合計
    s_cell.Offset(row_idx + 1, 0).Value = "合計"
    s_cell.Offset(row_idx + 1, 1).Formula = "=SUM(" + CStr(s_cell.Offset(2, 1).Address) + ":" + CStr(s_cell.Offset(row_idx, 1).Address) + ")"
    s_cell.Offset(row_idx + 1, 2).Formula = "=SUM(" + CStr(s_cell.Offset(2, 2).Address) + ":" + CStr(s_cell.Offset(row_idx, 2).Address) + ")"
    s_cell.Offset(row_idx + 1, 3).Formula = "=SUM(" + CStr(s_cell.Offset(2, 3).Address) + ":" + CStr(s_cell.Offset(row_idx, 3).Address) + ")"
    
    sh.Range(s_cell.Offset(1, 0), s_cell.Offset(1, 6)).Interior.ColorIndex = BClr_.Header
    sh.Range(s_cell.Offset(1, 0), s_cell.Offset(row_idx + 1, 6)).Borders.LineStyle = True
    sh.Range(s_cell.Offset(1, 4), s_cell.Offset(row_idx + 1, 6)).NumberFormatLocal = "0.0"

    ' ==============================================================
    ' 勝敗結果
    ' ==============================================================
    Set s_cell = sh.Range(GetAlp(COL_.entry_nm1) + CStr(row_max + 40))
    s_cell.Offset(1, 0).Value = "▼結果"
    
    s_cell.Offset(2, 0).Value = "○win_long"
    s_cell.Offset(3, 0).Value = "○win_short"
    s_cell.Offset(4, 0).Value = "●lose_long"
    s_cell.Offset(5, 0).Value = "●lose_short"
    
    
    alp1 = GetAlp(COL_.result)
    
    s_cell.Offset(1, 1).Value = "回数"
    s_cell.Offset(2, 1).Value = "=COUNTIF(" + alp1 + "2:" + alp1 + CStr(row_max) + ",""○win_long"")"
    s_cell.Offset(3, 1).Value = "=COUNTIF(" + alp1 + "2:" + alp1 + CStr(row_max) + ",""○win_short"")"
    s_cell.Offset(4, 1).Value = "=COUNTIF(" + alp1 + "2:" + alp1 + CStr(row_max) + ",""●lose_long"")"
    s_cell.Offset(5, 1).Value = "=COUNTIF(" + alp1 + "2:" + alp1 + CStr(row_max) + ",""●lose_short"")"
    
    s_cell.Offset(1, 2).Value = "発生確率"
    s_cell.Offset(2, 2).Value = "=" + CStr(s_cell.Offset(2, 1).Address) + "/SUM(" + CStr(s_cell.Offset(2, 1).Address) + ":" + CStr(s_cell.Offset(5, 1).Address) + ")"
    s_cell.Offset(3, 2).Value = "=" + CStr(s_cell.Offset(3, 1).Address) + "/SUM(" + CStr(s_cell.Offset(2, 1).Address) + ":" + CStr(s_cell.Offset(5, 1).Address) + ")"
    s_cell.Offset(4, 2).Value = "=" + CStr(s_cell.Offset(4, 1).Address) + "/SUM(" + CStr(s_cell.Offset(2, 1).Address) + ":" + CStr(s_cell.Offset(5, 1).Address) + ")"
    s_cell.Offset(5, 2).Value = "=" + CStr(s_cell.Offset(5, 1).Address) + "/SUM(" + CStr(s_cell.Offset(2, 1).Address) + ":" + CStr(s_cell.Offset(5, 1).Address) + ")"
    
    s_cell.Offset(1, 3).Value = "勝率"
    s_cell.Offset(2, 3).Value = "=" + CStr(s_cell.Offset(2, 1).Address) + "/(" + CStr(s_cell.Offset(2, 1).Address) + "+" + CStr(s_cell.Offset(4, 1).Address) + ")"
    s_cell.Offset(3, 3).Value = "=" + CStr(s_cell.Offset(3, 1).Address) + "/(" + CStr(s_cell.Offset(3, 1).Address) + "+" + CStr(s_cell.Offset(5, 1).Address) + ")"
    
    
    sh.Range(s_cell.Offset(1, 0), s_cell.Offset(1, 3)).Interior.ColorIndex = BClr_.Header
    sh.Range(s_cell.Offset(1, 0), s_cell.Offset(5, 3)).Borders.LineStyle = True
    sh.Range(s_cell.Offset(1, 2), s_cell.Offset(5, 3)).NumberFormatLocal = "0.00"

End Sub

Private Function GetAlp(col As Integer) As String
  Dim al As String
  al = Cells(1, col).Address(RowAbsolute:=False, ColumnAbsolute:=False)
  GetAlp = Left(al, Len(al) - 1)
End Function


